consumer.go

```diff
@@ -89,298 +145,453 @@ func NewConsumer(client *Client, topic string, partition int32, group string, co
+				continue
 			}
+
+			w.handleResponse(child, block)
 		}
 	}
 }
 
-func (c *Consumer) getOffset(where OffsetTime, retry bool) (int64, error) {
-	offset, err := c.client.GetOffset(c.topic, c.partition, where)
+func (w *consumerWorker) updateWorkCache(newWork []*PartitionConsumer) {
+	// take new work, and abandon work that has been closed
+	for _, child := range newWork {
+		w.work[child] = none{}
+	}
 
-	switch err {
-	case nil:
-		break
-	case EncodingError:
-		return -1, err
-	default:
-		if !retry {
-			return -1, err
+	for child, _ := range w.work {
+		select {
+		case <-child.dying:
+			close(child.trigger)
+			delete(w.work, child)
+		default:
 		}
+	}
+}
 
-		switch err.(type) {
-		case KError:
-			switch err {
-			case UnknownTopicOrPartition, NotLeaderForPartition, LeaderNotAvailable:
-				err = c.client.RefreshTopicMetadata(c.topic)
-				if err != nil {
-					return -1, err
-				}
-			default:
-				Logger.Printf("Unexpected error processing OffsetRequest; disconnecting broker %s: %s\n", c.broker.addr, err)
-				c.client.disconnectBroker(c.broker)
+func (w *consumerWorker) abort(err error) {
+	_ = w.broker.Close() // we don't care about the error this might return, we already have one
+	w.consumer.client.disconnectBroker(w.broker)
 
-				broker, brokerErr := c.client.Leader(c.topic, c.partition)
-				if brokerErr != nil {
-					return -1, brokerErr
-				}
-				c.broker = broker
-			}
-			return c.getOffset(where, false)
-		}
-		return -1, err
+	for child, _ := range w.work {
+		child.sendError(err)
+		child.trigger <- none{}
 	}
-	return offset, nil
-}
 
-// NewConsumerConfig creates a ConsumerConfig instance with sane defaults.
-func NewConsumerConfig() *ConsumerConfig {
-	return &ConsumerConfig{
-		DefaultFetchSize: 32768,
-		MinFetchSize:     1,
-		MaxWaitTime:      250 * time.Millisecond,
-		EventBufferSize:  16,
+	for newWork := range w.newWork {
+		for _, child := range newWork {
+			child.sendError(err)
+			child.trigger <- none{}
+		}
 	}
 }
 
-// Validate checks a ConsumerConfig instance. It will return a
-// ConfigurationError if the specified value doesn't make sense.
-func (config *ConsumerConfig) Validate() error {
-	if config.DefaultFetchSize <= 0 {
-		return ConfigurationError("Invalid DefaultFetchSize")
+func (w *consumerWorker) fetchNewMessages() (*FetchResponse, error) {
+	request := &FetchRequest{
+		MinBytes:    w.consumer.config.MinFetchSize,
+		MaxWaitTime: int32(w.consumer.config.MaxWaitTime / time.Millisecond),
 	}
 
-	if config.MinFetchSize <= 0 {
-		return ConfigurationError("Invalid MinFetchSize")
+	for child, _ := range w.work {
+		request.AddBlock(child.topic, child.partition, child.offset, child.fetchSize)
 	}
 
-	if config.MaxMessageSize < 0 {
-		return ConfigurationError("Invalid MaxMessageSize")
+	return w.broker.Fetch(w.consumer.client.id, request)
+}
+
+func (w *consumerWorker) handleResponse(child *PartitionConsumer, block *FetchResponseBlock) {
+	switch block.Err {
+	case NoError:
+		break
```

Comment:
I agree it is clearer this way. I just wanted to make sure I'm not missing something important.